# Notebook for making a exploratory data analysis 
    - 1ª Create descriptive statistics
    - 2ª Create business hypothesis
    - 3ª Testing business hypothesis

## Business problem:

The following statement is a fictitious problem:

Telco SA is a telecommunications company located in Canada, which lately has been trying problems, they **don't know what they know is with the company's measurements**, **we don't know how to take advantage and retain customers** or even identify the **customers' consumption patterns**. Your **current costs with ways of trying to retain customers are BRL 200,000.00**

**Key problem points to attack:**

**Problems with customers (Chur)**
- **Does not know the measures of the company**
- **We don't know how to leverage and retain customers**
- **Your current costs with customer campaigns of BRL 200.000.00**

**I will seek the solution of the problems described above in addition to trying to reduce these costs of Telco SA with Data Science techniques.**

**What is a company's churn?**

The **churn** is the percentage of customers who request the cancellation of the contracted service within a certain time interval.

**How ​​to calculate your business churn rate?**

The simplest formula to measure it is to divide the number of lost customers in a month by the number that there was no start of the month, and multiply by 100.

For example: we have 500 contacts, at the end of the month we identify 30 casualties, the churn rate will be 6%.

 - Dictionary of data
    - doc\Dictionary of Telco data.xlsx

# Libs

In [1]:
# open source
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns
import numpy as np
import sys

In [2]:
# configs
pd.set_option('display.max.columns', 50)

In [3]:
# my functions
# config path scripts
sys.path.insert(0, 'C:/Users/erico/OneDrive/Documentos/Projeto-Eric-Churn/Telco-Customer-Churn/scripts')

from extract_data import read_data

# Get data

In [4]:
# get dataset by Telco SA customer churn

# path to read specific csv data
path = r'C:\Users\erico\OneDrive\Documentos\Projeto-Eric-Churn\Telco-Customer-Churn\data\WA_Fn-UseC_-Telco-Customer-Churn.csv'

df = read_data(path)

df2 = df.copy()

In [5]:
# Print first rows of dataframe
df2.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0              Yes           Electronic check           29.85         29.85   
1               No               Mailed check           56.95       1889.50   
2              Yes               Mailed check           53.85        108.15   
3               No  Bank transfer (automatic)           42.30       1840.75   
4              Yes           Electronic check           70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

# Data overview

In [6]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [7]:
# check duplicateds
df2.duplicated().sum()

0

In [8]:
# check null fields
df2.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

 - So what for data sanity check ?
    - data not null
    - data not duplicated

# 1ª - Descriptive analysis
    - Make a statistic report from dataset

In [9]:
def make_stats_df(df:pd.DataFrame)-> pd.DataFrame:

    """ Function creates statistical report of an entire dataframe
        Receive:
            df - pandas dataframe with float or int type columns
        Returns:
            pandas dataframe and save Excel report in a data path
    """
    # filtra tipo de dados para int ou float antes da operação 
    df = df.select_dtypes(exclude=['object','datetime64'])

    # tendência central
    mean_ = pd.DataFrame(df.apply(np.mean)).T
    median_ = pd.DataFrame(df.apply(np.median)).T

    # dispersão e posição
    std_ = pd.DataFrame(df.apply( np.std ) ).T
    min_ = pd.DataFrame(df.apply( np.min )).T
    max_ = pd.DataFrame(df.apply( np.max )).T
    range_ = pd.DataFrame(df.apply(lambda x: x.max() - x.min() ) ).T
    skew_ = pd.DataFrame(df.apply(lambda x: x.skew() )).T
    kurtosis_ = pd.DataFrame(df.apply(lambda x: x.kurtosis() )).T          

    # concatenate     
    stats_df = pd.concat([min_, max_, range_, mean_, median_, std_, skew_, kurtosis_]).T.reset_index()
    stats_df.columns = ['Variáveis','Min','Max','Range','Mean','Median','Std','Skew','Kurtosis']

    return stats_df, stats_df.to_excel('Statistic_report_from_number_variables.xlsx',index = True), print('Report save in data path reports/ !')

 - Get general report statistics for numeric variables

In [10]:
df_number_report = make_stats_df(df2)

Report save in data path reports/ !


In [11]:
# Access dataframe report
df_number_report[0]

Variáveis    Min      Max   Range         Mean   Median          Std  \
0   SeniorCitizen   0.00     1.00     1.0     0.162147     0.00     0.368585   
1          tenure   0.00    72.00    72.0    32.371149    29.00    24.557737   
2  MonthlyCharges  18.25   118.75   100.5    64.761692    70.35    30.087911   
3    TotalCharges   0.00  8684.80  8684.8  2279.734304  1394.55  2266.633539   

       Skew  Kurtosis  
0  1.833633  1.362596  
1  0.239540 -1.387372  
2 -0.220524 -1.257260  
3  0.963235 -0.228580

 - So what dataframe number report:
    - not insigths know but on MonthlyCharges the range is $100.05 !

 - Get general statistics for categorics variables

In [12]:
df_categorical_report = df2.describe(exclude=['number']).T

In [13]:
df_categorical_report['pct'] = df_categorical_report['freq'] / df_categorical_report['count'] * 100

In [14]:
df_categorical_report

count unique               top  freq        pct
customerID        7043   7043        7590-VHVEG     1   0.014198
gender            7043      2              Male  3555   50.47565
Partner           7043      2                No  3641   51.69672
Dependents        7043      2                No  4933  70.041176
PhoneService      7043      2               Yes  6361  90.316626
MultipleLines     7043      3                No  3390  48.132898
InternetService   7043      3       Fiber optic  3096   43.95854
OnlineSecurity    7043      3                No  3498  49.666335
OnlineBackup      7043      3                No  3088  43.844952
DeviceProtection  7043      3                No  3095  43.944342
TechSupport       7043      3                No  3473  49.311373
StreamingTV       7043      3                No  2810  39.897771
StreamingMovies   7043      3                No  2785  39.542808
Contract          7043      3    Month-to-month  3875  55.019168
PaperlessBilling  7043      2               Yes  4171  59.221922
PaymentMethod     7043      4  Electronic check  2365  33.579441
Churn             7043      2                No  5174  73.463013

In [15]:
df_categorical_report.to_excel('Statistic_report_from_categorical_variables.xlsx', index=True)

 - So what categorical report 
    - Male predomine in 50.47% in the dataset Female 49,53% is balanced gender
    - Just over half of the base (51.69%) is not married
    - 70.04% of the base has no dependent
    - 90.31% of the base has telephone services
    - Less than half of the base has multiple phone lines (48.13%)
    - 43.95% Of the internet services are fiber optics
    - 49.66% of the customer don't have a OnlineSecurity
    - 43.84 of the customer don't have a OnlineBackup
    - 43.94% of the customer don't have a DeviceProtection
    - 49.31% of the customer don't have a TechSupport
    - 39.89% of the customer don't have a StreamingTV
    - 39.54% of the customer don't have a StreamingMovies
    - 55.01% the contract is Month-to-month
    - 59.22% of the payments is PaperlessBilling
    - 33.57% of the PaymentMethod is Electronic chec
    - 73.46% of the customer not is Churn